nemo/collections/asr/parts/submodules/conformer_modules.py

```diff
@@ -348,7 +348,7 @@ def forward(self, x, pad_mask=None, cache=None):
             x = self.pointwise_activation(x)
 
         if pad_mask is not None:
-            x = x.float().masked_fill(pad_mask.unsqueeze(1), 0.0)
+            x = x.masked_fill(pad_mask.unsqueeze(1), 0.0)
```

Comment:
Hmm can we cast to appropriate dtype explicitly. Vahid added this for a particular reason which I forget

Reply:
@VahidooX can you comment on why you added this cast? It is immediately followed by `self.depthwise_conv(x, cache=cache)`, which will end up casting the input `x` back down to fp16 when you run in automatic mixed precision. The only reason I can think of for why you might do this is if your cache is in float32, and adding a float16 value to the float32 cache caused a problem.

Reply:
Haven't heard back. Dropping this one.

Reply:
My bad, wasnt notified of ping. I am extremely wary of modifying any part of mha or relpos mha operations. I'd seriously like them to continue operating at fp32. Vahid worked a lot on this module, he has these explicit casts for a reason because he was unable to get stable training at all on fp16 and bf16 without these. 

The goal of this PR was to affect only inference, this affects training so id request to revert any changes to any file concerning MHA or RelPosMHA

Reply:
A middle ground is the following - check for self.training - if true, then use the explicit fp32 casts here and below, otherwise perform no cast (and we anyway explicitly cast to fp16/bf16 so it wont make a difference in inference)

Reply:
I sent @galv the details on this casting. It is not used for train or inference in nemo. @borisfom added that, I guess to make it work with trt or onxx conversions. Let's check it out with Boris before dropping that.
Here is the PR for this specific one:
https://github.com/NVIDIA/NeMo/pull/3787/files

Reply:
Sure if that works without the cast, do remove it - both ONNX exporter and TRT went very far since then.